python/docs/source/development/logger.rst

```diff
@@ -0,0 +1,149 @@
+
+- :meth:`PySparkLogger.log_info`
+- :meth:`PySparkLogger.log_warn`
+- :meth:`PySparkLogger.log_error`
+
+The logger can be easily configured to write logs to either the console or a specified file.
+
+Customizing Log Format
+======================
+The default log format is JSON, which includes the timestamp, log level, logger name, and the log message along with any additional context provided.
+
+Example log entry:
+
+.. code-block:: python
+
+    {
+      "timestamp": "2024-06-28T10:53:48.528Z",
+      "level": "ERROR",
+      "name": "DataFrameQueryContextLogger",
+      "message": "[DIVIDE_BY_ZERO] Division by zero.",
+      "context": {
+        "file": "/path/to/file.py",
+        "line_no": "17",
+        "fragment": "__truediv__"
+      },
+      "error_class": "DIVIDE_BY_ZERO"
+    }
+
+Setting Up
+==========
+To start using the PySpark logging module, you need to import the :class:`PySparkLogger` from the :ref:`pyspark.logger</reference/pyspark.logger.rst>`.
+
+.. code-block:: python
+
+    from pyspark.logger import PySparkLogger
+
+Usage
+=====
+Creating a Logger
+-----------------
+You can create a logger instance by calling the :meth:`PySparkLogger.get_logger`. By default, it creates a logger named "PySparkLogger" with an INFO log level.
+
+.. code-block:: python
+
+    logger = PySparkLogger.get_logger()
+
+Logging Messages
+----------------
+The logger provides three main methods for log messages: :meth:`PySparkLogger.log_info`, :meth:`PySparkLogger.log_warn`, and :meth:`PySparkLogger.log_error`.
+
+- **log_info**: Use this method to log informational messages.
+  
+  .. code-block:: python
+
+      user = "test_user"
+      action = "login"
+      logger.log_info(f"User {user} performed {action}", user=user, action=action)
+
+- **log_warn**: Use this method to log warning messages.
+  
+  .. code-block:: python
+
+      user = "test_user"
+      action = "access"
+      logger.log_warn("User {user} attempted an unauthorized {action}", user=user, action=action)
+
+- **log_error**: Use this method to log error messages.
+  
+  .. code-block:: python
+
+      user = "test_user"
+      action = "update_profile"
+      logger.log_error("An error occurred for user {user} during {action}", user=user, action=action)
+
+Logging to Console
+------------------
+
+.. code-block:: python
+
+    from pyspark.logger import PySparkLogger
+
+    # Create a logger that logs to console
+    logger = PySparkLogger.get_logger("ConsoleLogger")
+
+    user = "test_user"
+    action = "test_action"
+
+    logger.log_info(f"User {user} takes an {action}", user=user, action=action)
+
+This logs an information in the following JSON format:
+
+.. code-block:: python
+
+    {
+      "timestamp": "2024-06-28 19:44:19,030",
+      "level": "INFO",
+      "name": "ConsoleLogger",
+      "message": "User test_user takes an test_action",
+      "user": "test_user",
```

Comment:
In the scala version of structured logging, we put all the log variables into a map filed `context` so that we can have a simple table schema when loading the log files using `spark.read.json(...)`.

Reply:
If we put the variables in the top-level columns, it can make it hard to infer or specify the table schema. I would suggest that we keep the schema consistent.

Reply:
I agree that encapsulating all log variables within a "context" field would ensure schema consistency and simplify schema inference.

Let me update the log format to include a `context` field, containing all additional variables. The new log format will be:

```json
{
  "timestamp": "2024-06-28 19:44:19,030",
  "level": "INFO",
  "name": "ConsoleLogger",
  "message": "User test_user takes an test_action",
  "context": {
    "user": "test_user",
    "action": "test_action"
  }
}
```

Thanks for the suggestion!

Reply:
Updated. Thanks! :-)